fix: Add email verification requirement for API v1 and v2 email updates

anikdhabal opened the pull request:
## What does this PR do?

This PR fixes  email verification issue

**Changes:**

**API v2 (`/v2/me`)**:
- Checks if `email-verification` feature flag is enabled globally
- When changing to an unverified email: stores new email in `metadata.emailChangeWaitingForVerification` and sends verification email (primary email remains unchanged until verified)
- When changing to an already-verified secondary email: allows immediate change
- Adds `hasEmailBeenChanged` and `sendEmailVerification` response flags to indicate verification status
- Adds `findVerifiedSecondaryEmail` method to `UsersRepository`
- Adds `PrismaFeaturesRepository` to `MeModule` providers

**API v1 (`/v1/users/{userId}`)**:
- Implements same email verification logic as API v2
- Preserves existing response format (no new top-level fields)
- Uses `FeaturesRepository` to check feature flag

**Testing**:
- Adds E2E tests for API v2 verifying:
  - Email doesn't change immediately when verification is required
  - Metadata stores pending email correctly
  - Response flags indicate verification status
  - Verified secondary emails can be promoted immediately

devin-ai-integration[bot] committed:
fix: Add email verification requirement for API v1 and v2 email updates

- API v2 (/v2/me): Implement email verification check when updating primary email
  - Store new email in metadata as emailChangeWaitingForVerification when verification is required
  - Send verification email using sendChangeOfEmailVerification
  - Allow immediate email change if new email is already a verified secondary email
  - Add hasEmailBeenChanged and sendEmailVerification flags to response
  - Add tests to verify email verification behavior
  - Add findVerifiedSecondaryEmail method to UsersRepository
  - Add PrismaFeaturesRepository to MeModule providers

- API v1 (/v1/users/{userId}): Implement same email verification logic
  - Check if email-verification feature flag is enabled
  - Store new email in metadata when verification is required
  - Send verification email for unverified email changes
  - Allow immediate change for verified secondary emails
  - Preserve existing API v1 response format

- Test fixtures: Add createSecondaryEmail method to UserRepositoryFixture

This fixes the vulnerability where users could update their primary email
without verification via API endpoints.

Co-Authored-By: anik@cal.com <adhabal2002@gmail.com>

anikdhabal committed:
update

anikdhabal committed:
update

supalarry commented on apps/api/v2/src/ee/me/me.controller.ts:
```diff
@@ -60,18 +63,65 @@ export class MeController {
     @GetUser() user: UserWithProfile,
     @Body() bodySchedule: UpdateManagedUserInput
   ): Promise<UpdateMeOutput> {
+    const emailVerification = await this.featuresRepository.checkIfFeatureIsEnabledGlobally(
```
Controllers should only receive request, extract necessary data from requests like authenticated user and pass information to service and then service responds and controller's second responsibility is to then craft a response.

Given that, can we please create a new `me.service.ts` with a function `updateMe` which will all of the logic in this controller.

supalarry reviewed (changes requested):
Good stuff and great job on the tests! Have some thoughts:

devin-ai-integration[bot] committed:
refactor: Extract business logic to MeService and add platform-managed bypass

- Create MeService with all business logic extracted from MeController
- Update MeController to delegate to MeService (thin controller pattern)
- Add PrismaWorkerModule to MeModule imports to provide PrismaWriteService
- Add isPlatformManaged bypass: platform-managed users can update email without verification
- Fix test cleanup to use delete by ID instead of email to avoid failures when email changes
- Remove hasEmailBeenChanged and sendEmailVerification response flags per reviewer feedback
- Add comprehensive documentation to @ApiOperation describing email verification behavior
- Update tests to remove flag assertions

Addresses PR comments:
- supalarry: Extract logic to service layer
- supalarry: Allow platform-managed users to update email without verification
- supalarry: Remove response flags and document behavior in @ApiOperation instead
- cubic-dev-ai: Fix module dependency for PrismaFeaturesRepository
- cubic-dev-ai: Fix test cleanup issue

Co-Authored-By: anik@cal.com <adhabal2002@gmail.com>

devin-ai-integration[bot] committed:
docs: Simplify API operation description

Remove internal implementation details about metadata staging and shorten the description to focus on API consumer behavior.

Co-Authored-By: anik@cal.com <adhabal2002@gmail.com>

anikdhabal committed:
Merge branch 'main' into devin/api-v2-email-verification-1762510215

anikdhabal committed:
Merge branch 'main' into devin/api-v2-email-verification-1762510215

anikdhabal committed:
fix

anikdhabal committed:
Merge branch 'main' into devin/api-v2-email-verification-1762510215

anikdhabal committed:
update

supalarry commented on apps/api/v2/src/ee/me/services/me.service.ts:
```diff
@@ -0,0 +1,74 @@
+      }
+    }
+
+    const updatedUser = await this.usersRepository.update(user.id, update);
```
If we have unverified email that we delete `delete update.email`, then there is no need to run `this.usersRepository.update` - we should only run if the update data actually contains something.

supalarry reviewed (dismissed):
Love the tests and new service! Have some feedback on database calls and email changing logic:

anikdhabal committed:
update

anikdhabal commented on apps/api/v2/src/ee/me/services/me.service.ts:
```diff
@@ -0,0 +1,74 @@
+    const { user, updateData } = params;
+    const update = { ...updateData };
+
+    const emailVerification = await this.featuresRepository.checkIfFeatureIsEnabledGlobally(
```
Fixed it, thanks

anikdhabal commented on apps/api/v2/src/ee/me/services/me.service.ts:
```diff
@@ -0,0 +1,74 @@
+    let sendEmailVerification = false;
+
+    if (hasEmailBeenChanged && newEmail && !user.isPlatformManaged && emailVerification) {
+      const secondaryEmail = await this.usersRepository.findVerifiedSecondaryEmail(user.id, newEmail);
```
I don’t fully understand it, but one thing is missing. Before that, when the updated email is verified as the secondary email, we need to make the current primary email secondary and the secondary email primary

anikdhabal committed:
Merge branch 'main' into devin/api-v2-email-verification-1762510215

anikdhabal committed:
update

anikdhabal committed:
remove comment

anikdhabal committed:
Merge branch 'main' into devin/api-v2-email-verification-1762510215

Ryukemeister commented on apps/api/v1/pages/api/users/[userId]/_patch.ts:
```diff
@@ -130,10 +132,86 @@ export async function patchHandler(req: NextApiRequest) {
+
+    if (emailVerification) {
+      if (secondaryEmail && secondaryEmail.emailVerified) {
+        const [, updatedUser] = await prisma.$transaction([
```
I think we should move this over to a repository, similar to what you did in the `featuresRepository` above

anikdhabal committed:
addressed commit

anikdhabal committed:
Merge branch 'main' into devin/api-v2-email-verification-1762510215

supalarry commented on apps/api/v2/src/ee/me/services/me.service.ts:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    // Filter out undefined values and only update if there are actual fields to update
+    const filteredUpdate = Object.entries(update).reduce((acc, [key, value]) => {
```
Do we need this filtering? As far as i know if you try to submit any key as undefined and make request it will be 400:

<img width="1707" height="441" alt="Screenshot 2026-01-07 at 11 28 01" src="https://github.com/user-attachments/assets/618b0eb1-c2f9-408c-9dfc-3dcd1c6820d7" />

supalarry commented on apps/api/v2/src/ee/me/services/me.service.ts:
```diff
@@ -0,0 +1,115 @@
+        return {
+          updatedUser,
+        };
+      } else {
```
Since we have early return in the if statement `if (secondaryEmail && secondaryEmail.emailVerified) {` above, I would just get rid of else and put it on the same level as `const filteredUpdate = Object.entries(update).reduce((acc, [key, value]) => {` to reduce code nesting and simplify flow - either we enter if and return or we proceed with case B, but not it's harder to understand that IMO.

supalarry reviewed (commented):
Review:

anikdhabal committed:
review addressed

supalarry reviewed (dismissed):
Looks good!

anikdhabal commented on apps/api/v2/src/ee/me/services/me.service.ts:
```diff
@@ -0,0 +1,104 @@
+            where: { id: user.id },
+            data: update,
+          }),
+        ]);
```
I don't think a repository, but a service instead

anikdhabal committed:
use repository

anikdhabal committed:
Merge branch 'main' into devin/api-v2-email-verification-1762510215

anikdhabal committed:
Merge branch 'main' into devin/api-v2-email-verification-1762510215
